include/swift/Syntax/RawSyntax.h

```diff
@@ -218,43 +218,58 @@ class RawSyntax final
   }
 
   /// Constructor for creating layout nodes.
-  /// If the node has been allocated inside the bump allocator of a
-  /// \c SyntaxArena, that arena must be passed as \p Arena to retain the node's
-  /// underlying storage.
+  /// \p Children is an iterator that provides the child \c RawSyntax nodes.
+  /// It is only traversed once.
+  /// \p NumChildren must be equal to the number of elements provided by the \p
+  /// Children iterator.
   /// If \p NodeId is \c None, the next free NodeId is used, if it is passed,
   /// the caller needs to assure that the node ID has not been used yet.
-  RawSyntax(SyntaxKind Kind, ArrayRef<const RawSyntax *> Layout,
-            size_t TextLength, SourcePresence Presence,
+  template <typename ChildrenIteratorType>
+  RawSyntax(SyntaxKind Kind, ChildrenIteratorType Children,
+            uint32_t NumChildren, SourcePresence Presence,
```

Comment:
`ChildrenIteratorType` is not a correct name in the current implementation `ChildrenRangeType`.

But as I mentioned before, passing a range (a pair of start and end iterators) and the size is redundant. I think  the start iterator and the size are sufficient.
```cpp
    template <typename ChildrenIteratorType>
    RawSyntax(SyntaxKind Kind, ChildrenIteratorType ChildIter, uint32_t NumChildren,
              SourcePresence Presence, const RC<SyntaxArena> &Arena,
              llvm::Optional<SyntaxNodeId> NodeId) ... {

      const RawSyntax **TrailingChildren =
         getTrailingObjects<const RawSyntax *>();

      for (size_t numRemaining = NumChildren; numRemaining != 0; --numRemaining) {
        const RawSyntax *child = *ChildIter++;        
        *TrailingChildren++ = child;
        if (child) { ... }
      }
    }
```
In the user side,
```cpp
  OpaqueSyntaxNode
  SyntaxTreeCreator::recordRawSyntax(syntax::SyntaxKind kind,
                                    ArrayRef<OpaqueSyntaxNode> elements) {
   auto childIterator = reinterpret_cast<const RawSyntax *const *>(elements.data());
   auto raw = RawSyntax::make(kind, childIter, rawChildren.size(),
                              SourcePresence::Present, Arena);
  ...
```
```cpp
  OpaqueSyntaxNode SyntaxTreeCreator::makeDeferredLayout(
    syntax::SyntaxKind kind, bool IsMissing,
     const MutableArrayRef<ParsedRawSyntaxNode> &parsedChildren) {
   ...
   auto childIterator = llvm::map_iterator(
       parsedChildren.begin(),
       [](ParsedRawSyntaxNode &parsedChild) -> const RawSyntax * {
         return static_cast<const RawSyntax *>(parsedChild.takeData());
       });
   auto raw = RawSyntax::make(kind, childIterator, parsedChildren.size(),
   ...
```

Reply:
Sorry, I misunderstood you. I thought you were proposing not passing the size (which is necessary). Makes sense.